### DISCUSSION TITLE
Couldn’t verify the CA certificate as Budi is installed in docker container

### DISCUSSION THREAD

**Discussion by eivorwolfkissed** [Budibase/budibase#3917] [Help] [answered]
Dear Budibase Team,
We were testing budibase and built a couple of apps in it. The tool is intuitive to build web apps with little knowledge on front end programming.
As a next step, we tried to use OpenID Connect for user access management. Our IT helped us in installing certificate for this. However we face the following issue now:
FetchError: request to https://someurl/SAAS/auth/auth/.well-known/openid-configuration failed, reason: unable to verify the first certificate
    at ClientRequest.<anonymous> (/app/node_modules/node-fetch/lib/index.js:1461:11)
    at ClientRequest.emit (events.js:326:22)
    at ClientRequest.EventEmitter.emit (domain.js:483:12)
    at TLSSocket.socketErrorListener (_http_client.js:427:9)
    at TLSSocket.emit (events.js:314:20)
    at TLSSocket.EventEmitter.emit (domain.js:483:12)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  type: 'system',
  errno: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE',
  code: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE'
}
ERROR    [1641226478887] (29 on 5a436525954d): Error constructing OIDC authentication strategy
    Error: Error constructing OIDC authentication strategy
        at Object.exports.strategyFactory (/app/node_modules/@budibase/auth/src/middleware/passport/oidc.js:123:11)
        at runMicrotasks (<anonymous>)
        at processTicksAndRejections (internal/process/task_queues.js:97:5)

“The problem is because budi couldn’t verify the CA cert. Eventhough you installed the relevant cert on the host server, budi is running in a docker container, and you need to pass the CA cert path as env in the docker compose.”
Question:
Do you have any suggestion to overcome this error or way to pass the CA cert path as env in the docker compose ?

**Comment by mungki3** [comment]
I'd just like to echo that i've the same behaviour with mine too.

**Reply by JCV67** [reply]
Hi have a similar problem
https://github.com/Budibase/budibase/discussions/6825

**Comment by mjashanks** [comment]
Hi There, we currently use Envoy as a proxy, which comes packaged in the Budibase stack. You would need to make changes at the envoy level to do this.

However, because of this issue (and others) we have decided to swap Envoy for NGINX - this month. So this should be better supported at the end of this month.

**Reply by eivorwolfkissed** [reply]
Thank you for the suggestion @mjashanks. I will see if our team can make changes at the envoy level. Will be keen on your swap update !!

**Reply by sudpuzzer** [reply]
Since Budibase has changed over to NGINX for the proxy, is there any guidance available on this scenario? I have also set up an OIDC provider with a self-signed SSL certificate and would like very much to import my CA certificate into Budibase, but it's unclear why this is an NGINX problem and how to get around it.

**Reply by serhii-bielozorov** [reply]
Guys, are there any updates on this issue?

**Comment by sudpuzzer** [comment, marked as answer]
I managed to import a custom CA cert and make this work through the docker-compose file. First, I guessed that the budibase worker service is what facilitates the OIDC auth. I looked at the dockerfile for the worker image and saw that it was based on a Node JS image, so I found an environment variable for Node to add a custom CA ([link](https://stackoverflow.com/questions/29283040/how-to-add-custom-certificate-authority-ca-to-nodejs)).

Here's how I made it work:

1. Place your custom CA cert in a directory accessible to docker, I placed mine in the same folder as my docker-compose.yaml file, for example on Ubuntu: `/home/docker/budibase/RootCA.crt`
2. Start from the official docker-compose file found [here](https://raw.githubusercontent.com/Budibase/budibase/master/hosting/docker-compose.yaml), locate the worker-service entry.
3. Add a new bind mount to the worker-service like the following:

> ...
> services:
> ...
> &nbsp;&nbsp;worker-service:
> &nbsp;&nbsp;...
> &nbsp;&nbsp;&nbsp;&nbsp;volumes:
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- /home/docker/budibase/RootCA.crt:/app/cacert.crt
> ...

Make sure to set the first half of the bind mount to the actual location where you put your CA cert file. In this case, since it's in the same directory as my docker-compose file, I could have used `./RootCA.crt:/app/cacert.crt`
4. Add the variable NODE_EXTRA_CA_CERTS to the worker-service like the following:

> ...
> services:
> ...
> &nbsp;&nbsp;worker-service:
> &nbsp;&nbsp;...
> &nbsp;&nbsp;&nbsp;&nbsp;environment:
> &nbsp;&nbsp;&nbsp;&nbsp;...
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;NODE_EXTRA_CA_CERTS: /app/cacert.crt
> ...

5. That's it! Budibase shouldn't complain anymore about a certificate signed by your custom CA when connecting to an OIDC provider such as Keycloak.

Here is the whole worker-service section of my docker-compose file for reference:

> &nbsp;&nbsp;worker-service:
> &nbsp;&nbsp;&nbsp;&nbsp;restart: always
> &nbsp;&nbsp;&nbsp;&nbsp;image: budibase.docker.scarf.sh/budibase/worker
> &nbsp;&nbsp;&nbsp;&nbsp;container_name: bbworker
> &nbsp;&nbsp;&nbsp;&nbsp;environment:
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SELF_HOSTED: 1
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;PORT: 4003
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;CLUSTER_PORT: ${MAIN_PORT}
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;JWT_SECRET: ${JWT_SECRET}
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;MINIO_URL: http://minio-service:9000
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;APPS_URL: http://app-service:4002
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;COUCH_DB_USERNAME: ${COUCH_DB_USER}
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;COUCH_DB_PASSWORD: ${COUCH_DB_PASSWORD}
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;COUCH_DB_URL: http://${COUCH_DB_USER}:${COUCH_DB_PASSWORD}@couchdb-service:5984
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SENTRY_DSN: https://a34ae347621946bf8acded18e5b7d4b8@o420233.ingest.sentry.io/5338131
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;INTERNAL_API_KEY: ${INTERNAL_API_KEY}
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;REDIS_URL: redis-service:6379
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;REDIS_PASSWORD: ${REDIS_PASSWORD}
> **&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;NODE_EXTRA_CA_CERTS: /app/cacert.crt**
> &nbsp;&nbsp;&nbsp;&nbsp;depends_on:
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- redis-service
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- minio-service
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- couch-init
> **&nbsp;&nbsp;&nbsp;&nbsp;volumes:
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- ./RootCA.crt:/app/cacert.crt**

**Reply by pandel** [reply]
Made my day! Many thanks!

**Comment by maximebakercssc** [comment]
It is the same configuration for single image mode?
